Initial New Project Wizard scaffolding

sharon-wang opened the pull request:
<!-- Thank you for submitting a Pull Request. Please:
* Read our Pull Request guidelines:
  https://connect.posit.it/positron-wiki/pull-requests.html
* Ensure that the code is up-to-date with the `main` branch.
-->

### Intent

- First PR for #2110.

This PR mainly includes scaffolding and initial setup/components for the New Project flow, currently only allowing for the creation of a new folder based on the user's project name and directory selections. At this moment, the usable functionality of the New Project wizard is equivalent to the New Folder modal.

This video shows the "happy path" flow:

https://github.com/posit-dev/positron/assets/25834218/b1970e65-9da4-44a1-bb08-6efb8a4bbbbc

This video shows the different ways to trigger the New Project Wizard, as well as what's stored in the `projectConfig` data structure upon "accepting" the wizard modal.

https://github.com/posit-dev/positron/assets/25834218/354902a7-4121-4a3d-bfb8-918fdfb42241

UI Notes
- checking the git repo checkbox doesn't yet do anything
- project type selection doesn't yet do anything
- radio buttons on the Python interpreter page don't yet do anything
- Venv/Conda dropdown options are hardcoded text
- Conda Python interpreter versions are hardcoded text

Dev Notes
- this is feature-flagged by adding `IsDevelopmentContext` to the precondition for the New Project Wizard action
- lots of TODOs in the code! I am keeping track of what still needs to be done/fixed this way
- lots of commented out code in the components as placeholders for now
- lots of strings are not localized yet

### Approach

- Create the modal for the New Project flow based on the New Folder modal
- Add the New Project action to the same places where you can create a new folder
    - Application menu bar: File > New Project…
    - IDE Actions bar: New > New Project…
    - IDE Top Action Bar Custom Folder Menu: {myCurrentFolderName} > New Project…
    - Command Palette (Cmd+Shift+P): Workspaces: New Project…
- Handle navigating between steps (back/next buttons) in the wizard
    - Create an action bar with ok, cancel, back and next buttons
    - Each step in the modal is called a WizardStep (e.g, setting the project name and location is a WizardStep, setting up the Python environment/interpreter is a WizardStep)
    - Each substep in a WizardStep is called a WizardSubStep (e.g., setting the project name is a substep, setting the project parent directory is a substep, choosing if the project should be initialized as a git repo is a substep)
- Store the user's project configuration state via `useContext`, so the project config can be accessed from any step
- Populate the Python interpreter dropdown with the detected runtimes from the `languageRuntimeService.registeredRuntimes`

Here's a visualization of the component hierarchy:
<img src="https://github.com/posit-dev/positron/assets/25834218/1001df8f-6a58-48fb-8d5b-35f64a7c0d2b" width="500px">
- The `CurrentStep` is replaced with the appropriate step each time the user navigates in the wizard (using Back/Next buttons)
   - Steps that currently exist: projectTypeStep, projectNameLocationStep, pythonEnvironmentStep

### QA Notes

At this point, this work is not meant to be QA'd.

sharon-wang committed:
add New Project to top action bar and custom folder menu

Functionality is a duplication of New Folder.

sharon-wang committed:
create OKCancelBackNextActionBar component

sharon-wang committed:
create OKCancelBackNextModalDialog component

sharon-wang committed:
WIP: new project modal dialog layout

sharon-wang committed:
use WizardStep and WizardSubStep to organize inputs

sharon-wang committed:
remove dependency install step and fix up some code text styling

sharon-wang committed:
update newProjectModalDialog with merged dialog enhancements

sharon-wang committed:
use positron Button instead of HTML button in OKCancelBackNextActionBar

sharon-wang committed:
get useContext somewhat working and separate out step components

sharon-wang committed:
fix circular dependency issues

Can't use keyboard 'Enter' key to create the project.

sharon-wang committed:
Update newProjectModalDialog with modal renderer changes

sharon-wang committed:
clean up NewProjectType options

sharon-wang committed:
display the detected python runtimes in the combobox

re-rendering the combo box with the detected python runtimes isn't working yet. Currently, you need to go back a step and come back to the python env step to see the populated combo box.

sharon-wang committed:
add initial placeholder for python environment type selection

sharon-wang committed:
get python interpreter dropdown rendering and filtering working

sharon-wang committed:
add feature flag around PositronNewProjectAction

While the new project feature is under development, include precondition `IsDevelopmentContext`

sharon-wang committed:
fix up New Project related CSS file formatting

sharon-wang committed:
set the default color for ComboBox title and items

sharon-wang committed:
remove `hide` property from `ActionBarButtonConfig`

If the button should be hidden, the config corresponding to the button should be supplied to the OKCancelBackNextActionBar.

sharon-wang committed:
add comments and note TODOs in the code

sharon-wang committed:
fix merge conflicts after rebase with modal dialog refactor

sharon-wang committed:
add jsdoc for python interpreter drop down utils

sharon-wang committed:
remove extra leading separator in python interpreters dropdown

sharon-wang committed:
fix scoping of NewProjectWizardContextProvider

Resolves the issue where the Enter key can't be used to accept the modal.

sharon-wang committed:
alphabetize services for new project wizard

sharon-wang committed:
don't overwrite the project name if it's already set

...unless the project type has changed.

sharon-wang committed:
update React imports to use ESM import instead of CJS

sharon-wang committed:
fix filename for css import

softwarenerd reviewed (approved):
Looks good. Some localization missing where I mentioned it and elsewhere. 😀

nstrayer reviewed (approved):
Looks good to me! 
I had some React style comments. 

One thing I'm curious about is the file organization. it seems like most of the code is in a subfolder of `workbench/browser/positronModalDialogs`, is there a reason for this rather than having it in its own folder such as `workbench/browser/newProjectWizard`?

sharon-wang committed:
test that the new project wizard context is defined before using it

sharon-wang committed:
add missing string localizations or localization TODOs

sharon-wang committed:
use logService for logging errors instead of console.log

sharon-wang committed:
remove unused empty css file and fix minor css formatting

sharon-wang committed:
use ternary operator for conditional rendering

...instead of using the && operator, to avoid hitting unexpected behaviour when the object (left operand) being checked is falsy (e.g. 0, NaN), but when returned ends up being rendered as a string.

sharon-wang committed:
reorganize new project wizard directory structure

sharon-wang committed:
rename checkbox prop `defaultValue` to `initialChecked`
